frontend/catalyst/compilation_pipelines.py

```diff
@@ -194,7 +194,7 @@ def load_symbols(shared_object_file, func_name):
             setup: handle to the setup function, which initializes the device
             teardown: handle to the teardown function, which tears down the device
         """
-        shared_object = ctypes.CDLL(shared_object_file)
+        shared_object = ctypes.PyDLL(shared_object_file)
```

Comment:
@erick-xanadu I see the benefits of handling GIL automatically for the calls into the shared obj file by using `PyDLL` instead of `CDLL` but my main concern here is with the performance. My brief experience with `PyDLL` in the past thought me the consideration of not using it for the sake of efficiency due to some known reasons. So, I'm now wondering about the side effects of this change in our pipeline. Some benchmarks would be really nice.

I know that there are ways we can manually control GIL carefully to keep the program thread-safe and I want to know if this suggestion would help us of not using `PyDLL`. 

PS. I just saw your timing results; I'm not sure this is closely related to my points here but I'm glad you have generated those results and we can definitely discuss it... :100:

Reply:
The use of PyDLL is about being able to call into the Python C-API, I don't think we care much about the GIL. (@erick-xanadu can correct me here)

> My brief experience with PyDLL in the past thought me the consideration of not using it for the sake of efficiency due to some known reasons.

Given the benchmarks above, do you mind sharing these reasons if you any more insight?

Reply:
> The use of PyDLL is about being able to call into the Python C-API, I don't think we care much about the GIL. (@erick-xanadu can correct me here)

That's correct. As a consequence of calling the Python C-API from C code we must hold the GIL. This is unavoidable as the call calls to the Python C-API interact with the Python VM.

> Given the benchmarks above, do you mind sharing these reasons if you any more insight?

I suspected that having an additional call to the another compiler would increase compilation time (`clang++` or `cc`, it works with either), but without profiling it is impossible to know just how much. (The other alternative I had in mind uses `libffi` as a build time dependency which to me seemed like a potentially bad trade-off. Why do we need this dependency if we can JIT?) And it seemed a good question to at least know what is the cost of this JIT-ing vs adding a new build time dependency.

However, both holding the GIL and adding a new compilation step contribute to the problem. I believe holding the GIL reduces the throughput at which pytest goes through all of the different tests while adding a new compilation step increases the latency of each individual compilation. I did some simple experiments on Thursday where I stopped holding the GIL and there was some noticeable reduction when running `time pytest ...` but it was not enough to blame it on holding the GIL.

There is even one more alternative which would be to perform an LLVM-IR transformation in the generated code to create a wrapper around the `_mlir_ciface` function which will catch an exception. Here are some [instructions](https://llvm.org/docs/LangRef.html#landingpad-instruction). There is some recent discussion on also improving the LLVM MLIR dialect for exception related operations [here](https://discourse.llvm.org/t/improvements-to-llvm-dialect-exception-related-operations/69503).

While I think that both `libffi` and generating the code to handle the exceptions in LLVM-IR is sufficient for this case, I am also trying to think just how much more extensibility do we need for the memory issue.

Moving the memory from C to Python definitely requires the python C-API, but it also requires knowing the layout of the return value. The layout of the return value depends on the function generated. We can derive the layout if we assume that we will always return a struct of tensors (which we do at the moment) but that got me thinking about whether it makes sense to now generalize our struct of ranked tensors into a struct of unranked tensors, as this would be more future proofing and is more general at a small increase in runtime (unranked tensors are just a struct containing a field which correspond to the rank and a void* corresponding to the ranked tensor implementation).